In [1]:
import pandas as pd
import numpy as np
import os
import yaml


In [3]:
BASE = "/content/ecopackai"

DATA_PATH = f"{BASE}/materials_engineered.csv"
COST_PRED_PATH = f"{BASE}/rf_cost_metrics.csv"
CO2_PRED_PATH = f"{BASE}/co2_metrics.csv"

OUTPUT_DIR = f"{BASE}/outputs"
CONFIG_DIR = f"{BASE}/config"
DOCS_DIR = f"{BASE}/docs"
SRC_DIR = f"{BASE}/src/recommendation"

os.makedirs(OUTPUT_DIR, exist_ok=True)
os.makedirs(CONFIG_DIR, exist_ok=True)
os.makedirs(DOCS_DIR, exist_ok=True)
os.makedirs(SRC_DIR, exist_ok=True)


In [4]:
df = pd.read_csv(DATA_PATH)
df.head()


Material ID                            Packaging Type    Material Type  \
0    MAT_0001                           Cardboard Boxes        Cardboard   
1    MAT_0002  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   
2    MAT_0003                  Steel Racks & Containers            Steel   
3    MAT_0004  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   
4    MAT_0005  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   

                         Suitable Product Categories  \
0  E-commerce, Food & Beverage, Consumer Goods, A...   
1  Fragile Items, Cosmetics, Pharmaceuticals, Int...   
2   Heavy Industrial Components, High-Security Goods   
3  Fragile Items, Cosmetics, Pharmaceuticals, Int...   
4  Fragile Items, Cosmetics, Pharmaceuticals, Int...   

                     Recommended Packaging Use Cases Supplier Region  \
0  Last-mile delivery and primary e-commerce pack...            EMEA   
1      Void-fill and cushioning for fragile products            APAC   
2  Secure, high-load international shipping and l...        AMERICAS   
3      Void-fill and cushioning for fragile products             ROW   
4      Void-fill and cushioning for fragile products           LATAM   

   Recyclability (%) Recyclability Category  Recycled Content (%)  \
0                 98                   High                  79.0   
1                100                   High                  93.0   
2                 87                   High                  78.0   
3                100                   High                  89.0   
4                100                   High                  92.0   

   Reusability (%)  ...  Total Material Weight (tons)  \
0             49.0  ...                         790.0   
1             31.0  ...                         545.0   
2            100.0  ...                        4979.0   
3             31.0  ...                         656.0   
4             28.0  ...                         549.0   

   Supplier Sustainability Compliance (%)  Weight per Unit (g)  \
0                                    85.0          7801.698598   
1                                    88.0          7368.748394   
2                                    88.0        339886.681685   
3                                    81.0          8742.470281   
4                                    81.0          7932.494329   

   Cost per kg (USD)  CO2_Impact_Index_raw  CO2_Impact_Index  \
0           0.287117              0.215183             21.52   
1           0.246989              0.147690             14.77   
2           0.073554              0.870000             87.00   
3           0.277953              0.146007             14.60   
4           0.213048              0.164028             16.40   

   Cost_Efficiency_Index_raw  Cost_Efficiency_Index  \
0                   0.704681                  70.47   
1                   0.714471                  71.45   
2                   0.174145                  17.41   
3                   0.700252                  70.03   
4                   0.717501                  71.75   

   Material_Suitability_Score_raw  Material_Suitability_Score  
0                        0.440278                       44.03  
1                        0.253472                       25.35  
2                        0.774306                       77.43  
3                        0.162500                       16.25  
4                        0.225694                       22.57  

[5 rows x 31 columns]

In [5]:
cost_pred = pd.read_csv(COST_PRED_PATH)
co2_pred = pd.read_csv(CO2_PRED_PATH)

df = df.merge(cost_pred, on=["product_id","material_id"])
df = df.merge(co2_pred, on=["product_id","material_id"])


KeyError: 'product_id'

In [6]:
print(df.columns)

Index(['Material ID', 'Packaging Type', 'Material Type',
       'Suitable Product Categories', 'Recommended Packaging Use Cases',
       'Supplier Region', 'Recyclability (%)', 'Recyclability Category',
       'Recycled Content (%)', 'Reusability (%)', 'Biodegradation Time (days)',
       'End-of-Life Disposal (%)', 'Carbon Footprint (kg CO2/unit)',
       'CO2 Emission per kg (estimated)', 'Waste Reduction Impact (%)',
       'Sustainability Target Progress (%)', 'Load Handling Score',
       'Moisture Resistance Score', 'Thermal Resistance Score',
       'Cost per Unit (USD)', 'Annual Usage (units)',
       'Total Material Weight (tons)',
       'Supplier Sustainability Compliance (%)', 'Weight per Unit (g)',
       'Cost per kg (USD)', 'CO2_Impact_Index_raw', 'CO2_Impact_Index',
       'Cost_Efficiency_Index_raw', 'Cost_Efficiency_Index',
       'Material_Suitability_Score_raw', 'Material_Suitability_Score'],
      dtype='object')


# Task
I have analyzed the DataFrame `df` and identified potential columns that could serve as 'predicted_cost' and 'predicted_co2'.

For 'predicted_cost', suitable columns include:
- `'Cost per kg (USD)'`
- `'Cost per Unit (USD)'`
- `'Cost_Efficiency_Index'`

For 'predicted_co2', suitable columns include:
- `'CO2 Emission per kg (estimated)'`
- `'Carbon Footprint (kg CO2/unit)'`
- `'CO2_Impact_Index'`

The dataframes `cost_pred` and `co2_pred` contain model performance metrics, not the actual predictions to be merged with `df`.

Please specify which of these existing columns you would like to use as 'predicted_cost' and 'predicted_co2' respectively. Alternatively, if you wish to generate new predictions using a model, please let me know.

## Analyze df for cost-related columns

### Subtask:
Examine the `df` DataFrame to identify existing columns that might represent 'predicted_cost', such as 'Cost per kg (USD)' or 'Cost_Efficiency_Index'.


**Reasoning**:
I need to examine the specified cost-related columns to understand their data and determine which one is most suitable for 'predicted_cost'. I will use `df.head()` and `df.describe()` to inspect their contents and statistics.



In [7]:
print("\n--- 'Cost per kg (USD)' ---")
print(df['Cost per kg (USD)'].head())
print(df['Cost per kg (USD)'].describe())

print("\n--- 'Cost per Unit (USD)' ---")
print(df['Cost per Unit (USD)'].head())
print(df['Cost per Unit (USD)'].describe())

print("\n--- 'Cost_Efficiency_Index' ---")
print(df['Cost_Efficiency_Index'].head())
print(df['Cost_Efficiency_Index'].describe())


--- 'Cost per kg (USD)' ---
0    0.287117
1    0.246989
2    0.073554
3    0.277953
4    0.213048
Name: Cost per kg (USD), dtype: float64
count    403.000000
mean       0.465242
std        0.217368
min        0.030250
25%        0.310380
50%        0.413834
75%        0.690074
max        0.839078
Name: Cost per kg (USD), dtype: float64

--- 'Cost per Unit (USD)' ---
0     2.24
1     1.82
2    25.00
3     2.43
4     1.69
Name: Cost per Unit (USD), dtype: float64
count    403.000000
mean       8.935931
std        7.092944
min        0.210000
25%        2.810000
50%        3.870000
75%       14.650000
max       25.950000
Name: Cost per Unit (USD), dtype: float64

--- 'Cost_Efficiency_Index' ---
0    70.47
1    71.45
2    17.41
3    70.03
4    71.75
Name: Cost_Efficiency_Index, dtype: float64
count    404.00000
mean      54.88901
std       16.52378
min       15.20000
25%       41.54000
50%       66.67000
75%       69.14000
max       75.20000
Name: Cost_Efficiency_Index, dtype: float64


**Reasoning**:
Now I need to examine the specified CO2-related columns to understand their data and determine which one is most suitable for 'predicted_co2'. I will use `df.head()` and `df.describe()` to inspect their contents and statistics.



In [8]:
print(
"--- 'CO2 Emission per kg (estimated)' ---")
print(df['CO2 Emission per kg (estimated)'].head())
print(df['CO2 Emission per kg (estimated)'].describe())

print(
"--- 'Carbon Footprint (kg CO2/unit)' ---")
print(df['Carbon Footprint (kg CO2/unit)'].head())
print(df['Carbon Footprint (kg CO2/unit)'].describe())

print(
"--- 'CO2_Impact_Index' ---")
print(df['CO2_Impact_Index'].head())
print(df['CO2_Impact_Index'].describe())

--- 'CO2 Emission per kg (estimated)' ---
0    0.54
1    0.32
2    3.25
3    0.31
4    0.31
Name: CO2 Emission per kg (estimated), dtype: float64
count    403.000000
mean       1.148387
std        0.720590
min        0.280000
25%        0.510000
50%        1.400000
75%        1.500000
max        3.250000
Name: CO2 Emission per kg (estimated), dtype: float64
--- 'Carbon Footprint (kg CO2/unit)' ---
0    0.78
1    0.52
2    3.67
3    0.48
4    0.41
Name: Carbon Footprint (kg CO2/unit), dtype: float64
count    403.000000
mean       1.186774
std        0.748728
min        0.310000
25%        0.780000
50%        0.990000
75%        1.180000
max        3.670000
Name: Carbon Footprint (kg CO2/unit), dtype: float64
--- 'CO2_Impact_Index' ---
0    21.52
1    14.77
2    87.00
3    14.60
4    16.40
Name: CO2_Impact_Index, dtype: float64
count    404.000000
mean      41.651757
std       20.269967
min       12.170000
25%       21.140000
50%       52.770000
75%       55.022500
max      100.000000
Na

Based on the analysis of the `df` DataFrame:

**For 'predicted_cost':**
- `'Cost per kg (USD)'`: Represents the cost of the material per kilogram. Values range from 0.03 to 0.84 USD.
- `'Cost per Unit (USD)'`: Represents the cost of the material per unit. Values range from 0.21 to 25.95 USD.
- `'Cost_Efficiency_Index'`: Appears to be a calculated index, ranging from 15.20 to 75.20, possibly indicating efficiency rather than a direct cost.

**For 'predicted_co2':**
- `'CO2 Emission per kg (estimated)'`: Represents estimated CO2 emissions per kilogram of material. Values range from 0.28 to 3.25 kg CO2/kg.
- `'Carbon Footprint (kg CO2/unit)'`: Represents the carbon footprint per unit of material. Values range from 0.31 to 3.67 kg CO2/unit.
- `'CO2_Impact_Index'`: Appears to be a calculated index, ranging from 12.17 to 100.00, likely indicating impact rather than a direct CO2 emission value.

The columns `'Cost per kg (USD)'`, `'Cost per Unit (USD)'`, `'CO2 Emission per kg (estimated)'`, and `'Carbon Footprint (kg CO2/unit)'` provide direct measures, while the 'Index' columns provide a scaled or aggregated score. Please specify which of these columns you would like to use for 'predicted_cost' and 'predicted_co2' respectively. For example:

- **'predicted_cost'**: `'Cost per kg (USD)'`
- **'predicted_co2'**: `'Carbon Footprint (kg CO2/unit)'`

## Propose column mapping or next steps

### Subtask:
Propose which existing columns in `df` could be used as 'predicted_cost' and 'predicted_co2', or outline the steps needed to generate these predictions if they are not already present and a model is intended to be used.


### Proposed Mapping for Predicted Values

Based on the analysis of the available columns and the understanding that direct measurements are generally preferred for 'predicted' values:

*   **'predicted_cost'** is proposed to be mapped to the **'Cost per kg (USD)'** column.
*   **'predicted_co2'** is proposed to be mapped to the **'CO2 Emission per kg (estimated)'** column.


**Please confirm if these proposed selections are correct.**

Alternatively, if you prefer different columns from the analyzed list for 'predicted_cost' or 'predicted_co2', please specify which ones.

Also, let me know if you would prefer to generate new predictions using a model instead.

## Map selected columns to 'predicted_cost' and 'predicted_co2'

### Subtask:
Assign the chosen columns ('Cost per kg (USD)' and 'CO2 Emission per kg (estimated)') to new columns named 'predicted_cost' and 'predicted_co2' in the DataFrame `df`.


**Reasoning**:
To fulfill the subtask, I will create two new columns, 'predicted_cost' and 'predicted_co2', by assigning values from 'Cost per kg (USD)' and 'CO2 Emission per kg (estimated)' respectively, and then display the head of the DataFrame to confirm the new columns.



In [9]:
df['predicted_cost'] = df['Cost per kg (USD)']
df['predicted_co2'] = df['CO2 Emission per kg (estimated)']
df.head()

Material ID                            Packaging Type    Material Type  \
0    MAT_0001                           Cardboard Boxes        Cardboard   
1    MAT_0002  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   
2    MAT_0003                  Steel Racks & Containers            Steel   
3    MAT_0004  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   
4    MAT_0005  Protective Fillers (Paper/Biodegradable)  Paper/Bio-Based   

                         Suitable Product Categories  \
0  E-commerce, Food & Beverage, Consumer Goods, A...   
1  Fragile Items, Cosmetics, Pharmaceuticals, Int...   
2   Heavy Industrial Components, High-Security Goods   
3  Fragile Items, Cosmetics, Pharmaceuticals, Int...   
4  Fragile Items, Cosmetics, Pharmaceuticals, Int...   

                     Recommended Packaging Use Cases Supplier Region  \
0  Last-mile delivery and primary e-commerce pack...            EMEA   
1      Void-fill and cushioning for fragile products            APAC   
2  Secure, high-load international shipping and l...        AMERICAS   
3      Void-fill and cushioning for fragile products             ROW   
4      Void-fill and cushioning for fragile products           LATAM   

   Recyclability (%) Recyclability Category  Recycled Content (%)  \
0                 98                   High                  79.0   
1                100                   High                  93.0   
2                 87                   High                  78.0   
3                100                   High                  89.0   
4                100                   High                  92.0   

   Reusability (%)  ...  Weight per Unit (g)  Cost per kg (USD)  \
0             49.0  ...          7801.698598           0.287117   
1             31.0  ...          7368.748394           0.246989   
2            100.0  ...        339886.681685           0.073554   
3             31.0  ...          8742.470281           0.277953   
4             28.0  ...          7932.494329           0.213048   

   CO2_Impact_Index_raw  CO2_Impact_Index  Cost_Efficiency_Index_raw  \
0              0.215183             21.52                   0.704681   
1              0.147690             14.77                   0.714471   
2              0.870000             87.00                   0.174145   
3              0.146007             14.60                   0.700252   
4              0.164028             16.40                   0.717501   

   Cost_Efficiency_Index  Material_Suitability_Score_raw  \
0                  70.47                        0.440278   
1                  71.45                        0.253472   
2                  17.41                        0.774306   
3                  70.03                        0.162500   
4                  71.75                        0.225694   

   Material_Suitability_Score  predicted_cost  predicted_co2  
0                       44.03        0.287117           0.54  
1                       25.35        0.246989           0.32  
2                       77.43        0.073554           3.25  
3                       16.25        0.277953           0.31  
4                       22.57        0.213048           0.31  

[5 rows x 33 columns]

In [11]:
cost_pred = pd.read_csv(COST_PRED_PATH)
co2_pred = pd.read_csv(CO2_PRED_PATH)

# Removed the merge operations as cost_pred and co2_pred do not contain merge keys for merging with df.

In [12]:
np.random.seed(42)
df["predicted_cost"] = np.random.uniform(5, 50, len(df))
df["predicted_co2"] = np.random.uniform(0.1, 5, len(df))


In [13]:
ranking_config = {
    "weights": {
        "cost": 0.3,
        "co2": 0.4,
        "suitability": 0.3
    },
    "constraints": {
        "min_recyclability": "B",
        "max_cost": 40,
        "min_suitability": 50
    },
    "top_n": 3
}


In [14]:
config_path = f"{CONFIG_DIR}/ranking_weights.yaml"
with open(config_path, "w") as f:
    yaml.dump(ranking_config, f)

config_path


'/content/ecopackai/config/ranking_weights.yaml'

In [15]:
def min_max(series, reverse=False):
    norm = (series - series.min()) / (series.max() - series.min())
    return 1 - norm if reverse else norm

df["cost_norm"] = min_max(df["predicted_cost"], reverse=True)
df["co2_norm"] = min_max(df["predicted_co2"], reverse=True)
df["suit_norm"] = min_max(df["Material_Suitability_Score"])


In [16]:
w = ranking_config["weights"]

df["ranking_score"] = (
    w["cost"] * df["cost_norm"] +
    w["co2"] * df["co2_norm"] +
    w["suitability"] * df["suit_norm"]
)


In [18]:
recy_map = {"A":4, "B":3, "C":2, "D":1}
min_recy = recy_map[ranking_config["constraints"]["min_recyclability"]]

df = df[
    (df["predicted_cost"] <= ranking_config["constraints"]["max_cost"]) &
    (df["Material_Suitability_Score"] >= ranking_config["constraints"]["min_suitability"]) &
    (df["Recyclability Category"].map(recy_map) >= min_recy)
]

In [20]:
df["rank"] = df.groupby("Material ID")["ranking_score"]\
                .rank(ascending=False, method="dense")

top_n = ranking_config["top_n"]

ranked = df[df["rank"] <= top_n]\
            .sort_values(["Material ID","rank"])

In [22]:
ranked[["Material ID","rank",
        "ranking_score","predicted_cost",
        "predicted_co2","Material_Suitability_Score"]].head()

Empty DataFrame
Columns: [Material ID, rank, ranking_score, predicted_cost, predicted_co2, Material_Suitability_Score]
Index: []

In [23]:
out_path = f"{OUTPUT_DIR}/material_rankings.csv"
ranked.to_csv(out_path, index=False)

out_path


'/content/ecopackai/outputs/material_rankings.csv'

In [24]:
ranker_code = '''
import pandas as pd
import yaml

def rank_materials(df, config_path):
    with open(config_path) as f:
        cfg = yaml.safe_load(f)

    w = cfg["weights"]
    c = cfg["constraints"]
    top_n = cfg["top_n"]

    def min_max(series, reverse=False):
        norm = (series - series.min()) / (series.max() - series.min())
        return 1 - norm if reverse else norm

    df = df.copy()
    df["cost_norm"] = min_max(df["predicted_cost"], reverse=True)
    df["co2_norm"] = min_max(df["predicted_co2"], reverse=True)
    df["suit_norm"] = min_max(df["Material_Suitability_Score"])

    df["ranking_score"] = (
        w["cost"] * df["cost_norm"] +
        w["co2"] * df["co2_norm"] +
        w["suitability"] * df["suit_norm"]
    )

    recy_map = {"A":4,"B":3,"C":2,"D":1}
    min_recy = recy_map[c["min_recyclability"]]

    df = df[
        (df["predicted_cost"] <= c["max_cost"]) &
        (df["Material_Suitability_Score"] >= c["min_suitability"]) &
        (df["Recyclability_Category"].map(recy_map) >= min_recy)
    ]

    df["rank"] = df.groupby("product_id")["ranking_score"]\
                    .rank(ascending=False, method="dense")

    return df[df["rank"] <= top_n].sort_values(["product_id","rank"])
'''


In [25]:
ranker_path = f"{SRC_DIR}/ranker.py"
with open(ranker_path, "w") as f:
    f.write(ranker_code)

ranker_path


'/content/ecopackai/src/recommendation/ranker.py'

In [26]:
doc_md = """
# Material Ranking Logic

## Objective
Rank candidate packaging materials for each product using predicted cost,
predicted CO₂ impact, and Material Suitability Score.

## Ranking Criteria
- Cost (lower is better)
- CO₂ emissions (lower is better)
- Suitability score (higher is better)

All metrics are normalized to 0–1 before combining.

## Composite Score
score = w_cost * cost_norm +
        w_co2 * co2_norm +
        w_suitability * suit_norm

Weights are configurable via YAML.

## Constraints
Materials are filtered if:
- Recyclability below threshold
- Cost exceeds ceiling
- Suitability below minimum

## Output
Top-N ranked materials per product with ranks and scores.
"""

doc_path = f"{DOCS_DIR}/material_ranking.md"
with open(doc_path, "w") as f:
    f.write(doc_md)

doc_path


'/content/ecopackai/docs/material_ranking.md'